src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,129 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		using var httpClient = new HttpClient();
+		GithubRelease? releaseData = await GithubRelease.GetLatestReleaseDataAsync();
+		Version? latestVersion = releaseData?.GetReleaseVersion();
+		if( latestVersion is null ) {
+			throw new BmxException( "Failed to grab the latest version" );
+		}
+
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+
+		string archiveName = GetOSFileName();
+		string? downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		long backupPathTimeStamp = DateTime.Now.Millisecond;
+		string backupPath = Path.Join( BmxPaths.OLD_BMX_VERSIONS_PATH, $"bmx-v{localVersion}-{backupPathTimeStamp}-old.bak" );
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
```

Comment:
IIUC
```suggestion
			throw new BmxException( "Could not remove the old version. Please try again with elevated permissions.", ex );
```

I know it's technically a move not a delete, but I think saying saying "move" would be more confusing to users not familiar with our implementation details.